### DISCUSSION TITLE
aes encryption performance esp32-c3 and esp32-s3

### DISCUSSION THREAD

**Discussion by puppet13th** [micropython/micropython#9401] [Hardware & Peripherals]
My lolin c3 mini and lolin s3 just arrived today. I'm curious about esp32-s3 performance compared to esp32-c3 and do aes encryption with these result:
lolin c3 mini:
```
MicroPython v1.19.1 on 2022-06-18; LOLIN_C3_MINI with ESP32-C3FH4
Type "help()" for more information.
>>> %run -c $EDITOR_CONTENT
1000 encrypt & decrypt iterations in 6910 ms
```
lolin s3(no spiram firmware):
```
MicroPython v1.19.1 on 2022-06-18; ESP32S3 module with ESP32S3
Type "help()" for more information.
>>> %run -c $EDITOR_CONTENT
1000 AES encrypt & decrypt iterations in 50909 ms
```
ESP32 d1-mini clone:
```
MicroPython v1.19.1 on 2022-09-24; ESP32 module with ESP32
Type "help()" for more information.
>>> %run -c $EDITOR_CONTENT
1000 encrypt & decrypt iterations in 9559 ms
```
all board running at base clock(160mhz).

rp2040:
```
MicroPython v1.19.1 on 2022-06-18; Raspberry Pi Pico with RP2040

Type "help()" for more information.

>>> %Run -c $EDITOR_CONTENT
1000 encrypt & decrypt iterations in 93854 ms
```
rp2040 placed last because lower frequency and does not have aes acceleration.

does esp32-s3 aes acceleration disabled in micropython?

**Comment by jimmo** [comment]
> rp2040 placed last because lower frequency and does not have aes acceleration.

You didn't show the code you used to test this, but assuming you're using https://docs.micropython.org/en/latest/library/cryptolib.html

In which case these all use mbedtls, and on ESP32* we used the default config from the ESP IDF.
See notes at https://github.com/espressif/esp-idf/blob/master/components/mbedtls/Kconfig#L353

**Reply by puppet13th** [reply]
example code:
```py
try:
    from ucryptolib import aes as AES
except:
    from Crypto.Cipher.AES import new as AES

from time import ticks_ms
from os import urandom
from machine import freq
freq(160_000_000)

for num in range (0,3):
    loop = 0
    ticks = ticks_ms()
    while loop < 1000:
        dummy = urandom(2048)
        key = urandom(32)
        aes = AES(key,1) # 256bit key cbc mode
        encrypted = aes.encrypt(dummy)
        aes = AES(key,1) # 256bit key cbc mode
        decrypted = aes.decrypt(encrypted)
        if dummy != decrypted:
            print('decrypt error!')
            break
        loop += 1
    print(num,ticks_ms() - ticks)
```

result ( pass , ms )

```
lolin c3-mini @160mhz
0 6127
1 6132
2 6133
lolin c3-mini @80mhz
0 11681
1 11697
2 11692
```

```
lolin-s3 spiram @160mhz
0 21172
1 21275
2 21277
lolin-s3 spiram @80mhz
0 30296
1 30541
2 30540
```

```
rp2040 @160mhz
0 22146
1 22144
2 22152
rp2040 @80mhz
0 35972
1 35967
2 35983
```

from those result i can make assumption:
1. look like rp2040 does not have aes acceleration ([datasheet](https://datasheets.raspberrypi.com/rp2040/rp2040-datasheet.pdf) does not mention aes)
2. esp32-s3 should have aes acceleration(according to Kconfig), but used in practice (have similar result to rp2040)
3. esp32-c3 uses aes acceleration

hopefully you can understand what i'm trying to show here.
